### PR TITLE
Fix libnghttp2 for shared=False and with_asio=True

### PR DESCRIPTION
**Pull request by xiaotianrandom**
Specify library name and version:  **libnghttp2/1.40.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

This patch fixes the following issues when building with `shared: False` and `with_asio: True`.
- The target `nghttp2_asio` should link the static target `nghttp2_static` instead of `nghttp2`.
- The target `nghttp2_asio` should be static as specified by the `shared` option.
- `nghttp2_static` should always be linked after `nghttp2_asio`.

I also add a new testcase `test_package_static_asio` and test it for both 1.39.2 and 1.40.0 (on Ubuntu 18.04 and conan 1.29.0)

Below is the details of the error without this PR. 

Running the new testcase gives the following error:
```
$ cd conan-center-index/recipes/libnghttp2/all
$ conan create . libnghttp2/1.40.0@ -tf test_package_static_asio
...
[ 98%] Linking CXX static library ../../lib/libnghttpx.a
[ 98%] Built target nghttpx_static
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
libnghttp2/1.40.0:
libnghttp2/1.40.0: ERROR: Package '699d4e95a2fb0c06debf97e065bdf4be5713bf36' build failed
libnghttp2/1.40.0: WARN: Build folder /home/github/.conan/data/libnghttp2/1.40.0/_/_/build/699d4e95a2fb0c06debf97e065bdf4be5713bf36
ERROR: libnghttp2/1.40.0: Error in build() method, line 111
        cmake.build()
        ConanException: Error 2 while executing cmake --build '/home/github/.conan/data/libnghttp2/1.40.0/_/_/build/699d4e95a2fb0c06debf97e065bdf4be5713bf36' '--' '-j8'
```

Re-running the erroneous command shows the real issue.
```
$ cmake --build '/home/github/.conan/data/libnghttp2/1.40.0/_/_/build/699d4e95a2fb0c06debf97e065bdf4be5713bf36'
[ 16%] Built target nghttp2_static
[ 17%] Built target url-parser
[ 20%] Built target llhttp
[ 21%] Linking CXX shared library ../../lib/libnghttp2_asio.so
/usr/bin/ld: cannot find -lnghttp2
collect2: error: ld returned 1 exit status
source_subfolder/src/CMakeFiles/nghttp2_asio.dir/build.make:829: recipe for target 'lib/libnghttp2_asio.so.1.0.0' failed
make[2]: *** [lib/libnghttp2_asio.so.1.0.0] Error 1
CMakeFiles/Makefile2:317: recipe for target 'source_subfolder/src/CMakeFiles/nghttp2_asio.dir/all' failed
make[1]: *** [source_subfolder/src/CMakeFiles/nghttp2_asio.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

**Reason**. The target `nghttp2_asio`, which is enabled by `with_asio: True`, links the target `nghttp2` in the original CMakeLists.txt ([here](https://github.com/nghttp2/nghttp2/blob/v1.40.0/src/CMakeLists.txt#L263)). With `shared: False`, it should link the target `nghttp2_static` instead. The original recipe already has a [patch](https://github.com/conan-io/conan-center-index/blob/4e4a37d511e7d0a42878e81e65e1bd3b34ad6855/recipes/libnghttp2/all/conanfile.py#L99-L106) to fix this, which is not enough. This PR adds two patches to fixes the `target_link_libraries(nghttp2_asio, ...)` and also changes the target `nghttp2_asio` to static when `shared: False`.

In addition, this PR also fixes the linking order to ensure that `nghttp2_static` is always linked after `nghttp2_asio`. The original recipe just uses `tools.collect_libs`, which seems to return libs in alphabetical order. It happens to work for nghttp2 1.40.0 as `nghttp2_static` is behind `nghttp2_asio`. But it fails for nghttp2 1.39.2 because output filename of the target `nghttp2_static` is changed to `nghttp2` [here](https://github.com/nghttp2/nghttp2/blob/v1.39.2/lib/CMakeLists.txt#L65) and thus comes before `nghttp2_asio` in alphabetical order.

### CODE REVIEW

**Comment by xiaotianrandom** [comment]
For testing, I created a new directory instead of modifying the existing one, because I thought it would be important to test the default option set. I'm just wondering if there's a way to tell the CI to run both testcases.

**Comment by uilianries** [comment]
@xiaotianrandom Could you please show the error which it will cause without this patch?

**Comment by xiaotianrandom** [comment]
Sure. Running the new testcase gives the following error:

```
$ cd conan-center-index/recipes/libnghttp2/all
$ conan create . libnghttp2/1.40.0@ -tf test_package_static_asio
...
[ 98%] Linking CXX static library ../../lib/libnghttpx.a
[ 98%] Built target nghttpx_static
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
libnghttp2/1.40.0:
libnghttp2/1.40.0: ERROR: Package '699d4e95a2fb0c06debf97e065bdf4be5713bf36' build failed
libnghttp2/1.40.0: WARN: Build folder /home/github/.conan/data/libnghttp2/1.40.0/_/_/build/699d4e95a2fb0c06debf97e065bdf4be5713bf36
ERROR: libnghttp2/1.40.0: Error in build() method, line 111
        cmake.build()
        ConanException: Error 2 while executing cmake --build '/home/github/.conan/data/libnghttp2/1.40.0/_/_/build/699d4e95a2fb0c06debf97e065bdf4be5713bf36' '--' '-j8'
```

Re-running the erroneous command shows the real issue.
```
$ cmake --build '/home/github/.conan/data/libnghttp2/1.40.0/_/_/build/699d4e95a2fb0c06debf97e065bdf4be5713bf36'[ 16%] Built target nghttp2_static
[ 17%] Built target url-parser
[ 20%] Built target llhttp
[ 21%] Linking CXX shared library ../../lib/libnghttp2_asio.so
/usr/bin/ld: cannot find -lnghttp2
collect2: error: ld returned 1 exit status
source_subfolder/src/CMakeFiles/nghttp2_asio.dir/build.make:829: recipe for target 'lib/libnghttp2_asio.so.1.0.0' failed
make[2]: *** [lib/libnghttp2_asio.so.1.0.0] Error 1
CMakeFiles/Makefile2:317: recipe for target 'source_subfolder/src/CMakeFiles/nghttp2_asio.dir/all' failed
make[1]: *** [source_subfolder/src/CMakeFiles/nghttp2_asio.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

Here is the reason. The target `nghttp2_asio`, which is enabled by `with_asio: True`, links to the target `nghttp2` in the original CMakeLists.txt ([here](https://github.com/nghttp2/nghttp2/blob/v1.40.0/src/CMakeLists.txt#L263)).

**Comment by xiaotianrandom** [comment]
Sure. I updated the original post and added details of the error there.
(Sorry for closing the issue by mistake just now.)

File: recipes/libnghttp2/all/test_package_static_asio/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.1)
```
Comment (uilianries) [inline recipes/libnghttp2/all/test_package_static_asio/CMakeLists.txt:1]: You must remove  recipes/libnghttp2/all/test_package_static_asio and its content. CCI won't run it.

**Review by uilianries — CHANGES_REQUESTED** [review]

File: recipes/libnghttp2/all/test_package_static_asio/CMakeLists.txt (hunk 1 above)
Comment (xiaotianrandom) [reply recipes/libnghttp2/all/test_package_static_asio/CMakeLists.txt]: Got it. I removed it.

**Review by uilianries — CHANGES_REQUESTED** [review]

File: recipes/libnghttp2/all/conanfile.py [hunk 2]
```diff
@@ -121,6 +134,14 @@ def package(self):
...
     def package_info(self):
         self.cpp_info.libs = tools.collect_libs(self)
+        # Make sure nghttp2 is linked after other libs when shared is False.
+        # Name of the static lib is "nghttp2" in 1.39.2 and "nghttp2_static" in 1.40.0.
+        if "nghttp2" in self.cpp_info.libs:
+            self.cpp_info.libs.remove("nghttp2")
+            self.cpp_info.libs.append("nghttp2")
+        if "nghttp2_static" in self.cpp_info.libs:
+            self.cpp_info.libs.remove("nghttp2_static")
+            self.cpp_info.libs.append("nghttp2_static")
```
[the remark this answers is not in the document]
Comment (xiaotianrandom) [reply recipes/libnghttp2/all/conanfile.py]: Thanks for the suggestion and I've updated the code. We do need to call `tools.collect_libs`. It returns `["nghttp2" + suffix]` when `with_asio: False`, and `["nghttp2" + suffix, "nghttp2_asio"]` when `with_asio: True`.

[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/libnghttp2/all/conanfile.py]: my point is, be explicit in this case, as the order is important. You can discard collect_libs and assign the list directly.

[the remark this answers is not in the document]
Comment (xiaotianrandom) [reply recipes/libnghttp2/all/conanfile.py]: I see. I updated the code again and discarded collect_libs.

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by intelligide — APPROVED** [review]

**Pull request merged**